utilities/src/main/java/com/splicemachine/utils/StringUtils.java

```diff
@@ -96,16 +107,54 @@ private static String padding(int repeat, char padChar) throws IndexOutOfBoundsE
         }
     }
 
+    public static String maskMessage(String message, Pattern maskPattern, String maskString) {
+        Matcher matcher = maskPattern.matcher(message);
+
+        List<Pair<Integer, Integer>> groups = new ArrayList<>();
+        while (matcher.find()) {
+            for (int i = 1; i <= matcher.groupCount(); i++) {
+                if (matcher.group(i) != null)
+                    groups.add(new Pair<>(matcher.start(i), matcher.end(i)));
+            }
+        }
+        if (groups.size() == 0) {
+            return message;
+        }
+        StringBuilder maskedMessage = new StringBuilder("");
+        for (int i = 0; i <= groups.size(); i++) {
+            int start;
+            int end;
+            if (i == 0) {
+                start = 0;
+            } else {
+                start = groups.get(i-1).getSecond();
+                maskedMessage.append(maskString);
+            }
+            if (i == groups.size()) {
+                end = message.length() - 1;
+            } else {
+                end = groups.get(i).getFirst() - 1;
+            }
+            maskedMessage.append(message, start, end+1);
+        }
+        return String.valueOf(maskedMessage);
+    }
+
+
     public static String logSpliceAuditEvent(String userid, String event, boolean status, String addr, String statement, String reason)
     {
+
+        Pattern maskPattern = Pattern.compile(maskPasswordPatternStr,Pattern.CASE_INSENSITIVE);
```

Comment:
Is it expensive to create and compile a pattern each time we log a entry? Since it is the same pattern that we are using, is it possible to let each thread share the same Pattern object?

Reply:
We can move the maskPattern out of the function to be a public static member of StringUtils class, the pattern won't be compiled each time and could be shared among threads.